external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -445,197 +529,68 @@ private[kafka010] case class InternalKafkaConsumer(
+      body
   }
 }
 
-
 private[kafka010] object KafkaDataConsumer extends Logging {
+  val UNKNOWN_OFFSET = -2L
 
   case class AvailableOffsetRange(earliest: Long, latest: Long)
 
-  private case class CachedKafkaDataConsumer(internalConsumer: InternalKafkaConsumer)
-    extends KafkaDataConsumer {
-    assert(internalConsumer.inUse) // make sure this has been set to true
-    override def release(): Unit = { KafkaDataConsumer.release(internalConsumer) }
-  }
-
-  private case class NonCachedKafkaDataConsumer(internalConsumer: InternalKafkaConsumer)
-    extends KafkaDataConsumer {
-    override def release(): Unit = { internalConsumer.close() }
-  }
-
-  private case class CacheKey(groupId: String, topicPartition: TopicPartition) {
+  case class CacheKey(groupId: String, topicPartition: TopicPartition) {
     def this(topicPartition: TopicPartition, kafkaParams: ju.Map[String, Object]) =
       this(kafkaParams.get(ConsumerConfig.GROUP_ID_CONFIG).asInstanceOf[String], topicPartition)
   }
 
-  // This cache has the following important properties.
-  // - We make a best-effort attempt to maintain the max size of the cache as configured capacity.
-  //   The capacity is not guaranteed to be maintained, especially when there are more active
-  //   tasks simultaneously using consumers than the capacity.
-  private lazy val cache = {
-    val conf = SparkEnv.get.conf
-    val capacity = conf.get(CONSUMER_CACHE_CAPACITY)
-    new ju.LinkedHashMap[CacheKey, InternalKafkaConsumer](capacity, 0.75f, true) {
-      override def removeEldestEntry(
-        entry: ju.Map.Entry[CacheKey, InternalKafkaConsumer]): Boolean = {
-
-        // Try to remove the least-used entry if its currently not in use.
-        //
-        // If you cannot remove it, then the cache will keep growing. In the worst case,
-        // the cache will grow to the max number of concurrent tasks that can run in the executor,
-        // (that is, number of tasks slots) after which it will never reduce. This is unlikely to
-        // be a serious problem because an executor with more than 64 (default) tasks slots is
-        // likely running on a beefy machine that can handle a large number of simultaneously
-        // active consumers.
-
-        if (!entry.getValue.inUse && this.size > capacity) {
-          logWarning(
-            s"KafkaConsumer cache hitting max capacity of $capacity, " +
-              s"removing consumer for ${entry.getKey}")
-          try {
-            entry.getValue.close()
-          } catch {
-            case e: SparkException =>
-              logError(s"Error closing earliest Kafka consumer for ${entry.getKey}", e)
-          }
-          true
-        } else {
-          false
-        }
-      }
+  private val consumerPool = InternalKafkaConsumerPool.build
+  private val fetchedDataPool = FetchedDataPool.build
+
+  ShutdownHookManager.addShutdownHook { () =>
```

Comment:
Hmm... probably doesn't matter much, but the use of shutdown hooks is sort of a "code smell".

e.g. if you have two Spark applications running in the same JVM, one after another, the second one will see the cached bits from the first one.

if possible it would be better to tie this to the lifecycle of the application itself (`SparkContext` instance), although I'm not sure how easy that is.

Reply:
I think the real issue is using `object` in some instance which is bound to the specific task, specific query, etc (if we consider multiple applications running in same JVM, then specific application too). I have seen other cases for this as well, and without extreme care, it causes an issue. e.g. #24946 

I'm also not sure Spark has some sort of lifecycle management and hook to handle such case, but I'll consider it. Maybe shutdown hook is still needed if Spark can't guarantee to signal listener for its lifecycle (like why Java should need shutdown hook).